pkgs/build-support/build-composer/default.nix

```diff
@@ -0,0 +1,222 @@
+# This file originates from composer2nix
```

Comment:
Can the changes be ported to composer2nix? Can it be imported somehow? What does @svanderburg have to say?

Reply:
I believe they can be backported yes.

The biggest change I've done is that I removed composer from this file because it pulled in it's own composer.

Then I did some minor changes to the PHP to be a bit cleaner and so.

Here's a diff from a generated composer-env.nix file and this file:
```diff
diff --git a/pkgs/build-support/build-composer/default.nix b/pkgs/build-support/build-composer/default.nix
index e0b82acfc5a..955a359595c 100644
--- a/pkgs/build-support/build-composer/default.nix
+++ b/pkgs/build-support/build-composer/default.nix
@@ -1,8 +1,41 @@
 # This file originates from composer2nix
 
-{ stdenv, makeWrapper, writeTextFile, fetchurl, php, unzip, composer }:
+{ stdenv, writeTextFile, fetchurl, php, unzip }:
 
 let
+  composer = stdenv.mkDerivation {
+    name = "composer-1.6.5";
+    src = fetchurl {
+      url = https://github.com/composer/composer/releases/download/1.6.5/composer.phar;
+      sha256 = "07xkpg9y1dd4s33y3cbf7r5fphpgc39mpm066a8m9y4ffsf539f0";
+    };
+    buildInputs = [ php ];
+
+    # We must wrap the composer.phar because of the impure shebang.
+    # We cannot use patchShebangs because the executable verifies its own integrity and will detect that somebody has tampered with it.
+
+    buildCommand = ''
+      # Copy phar file
+      mkdir -p $out/share/php
+      cp $src $out/share/php/composer.phar
+      chmod 755 $out/share/php/composer.phar
+
+      # Create wrapper executable
+      mkdir -p $out/bin
+      cat > $out/bin/composer <<EOF
+      #! ${stdenv.shell} -e
+      exec ${php}/bin/php $out/share/php/composer.phar "\$@"
+      EOF
+      chmod +x $out/bin/composer
+    '';
+    meta = {
+      description = "Dependency Manager for PHP";
+      #license = stdenv.licenses.mit;
+      maintainers = [ stdenv.lib.maintainers.sander ];
+      platforms = stdenv.lib.platforms.unix;
+    };
+  };
+
   buildZipPackage = { name, src }:
     stdenv.mkDerivation {
       inherit name src;
@@ -17,7 +50,7 @@ let
     };
 
   buildPackage =
-    { pname
+    { name
     , src
     , packages ? {}
     , devPackages ? {}
@@ -38,28 +71,36 @@ let
         text = ''
           #! ${php}/bin/php
           <?php
-          if (file_exists($argv[1])) {
+          if(file_exists($argv[1]))
+          {
               $composerLockStr = file_get_contents($argv[1]);
 
-              if ($composerLockStr === false) {
+              if($composerLockStr === false)
+              {
                   fwrite(STDERR, "Cannot open composer.lock contents\n");
                   exit(1);
               }
+              else
+              {
+                  $config = json_decode($composerLockStr, true);
 
-              $config = json_decode($composerLockStr, true);
+                  if(array_key_exists("packages", $config))
+                      $allPackages = $config["packages"];
+                  else
+                      $allPackages = array();
 
-              $allPackages = array_key_exists("packages", $config) ? $config["packages"] : array();
+                  ${stdenv.lib.optionalString (!noDev) ''
+                    if(array_key_exists("packages-dev", $config))
+                        $allPackages = array_merge($allPackages, $config["packages-dev"]);
+                  ''}
 
-              ${stdenv.lib.optionalString (!noDev) ''
-              if (array_key_exists("packages-dev", $config)) {
-                  $allPackages = array_merge($allPackages, $config["packages-dev"]);
+                  $packagesStr = json_encode($allPackages, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
+                  print($packagesStr);
               }
-              ''}
-
-              echo json_encode($allPackages, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
-          } else {
-              echo "[]";
           }
+          else
+              print("[]");
+          ?>
         '';
       };
 
@@ -71,24 +112,30 @@ let
           <?php
           $composerJSONStr = file_get_contents($argv[1]);
 
-          if ($composerJSONStr === false) {
-              fwrite(STDERR, 'Cannot open composer.json contents'.PHP_EOL);
+          if($composerJSONStr === false)
+          {
+              fwrite(STDERR, "Cannot open composer.json contents\n");
               exit(1);
           }
+          else
+          {
+              $config = json_decode($composerJSONStr, true);
 
-          $config = json_decode($composerJSONStr, true);
-
-          $binDir = array_key_exists('bin-dir', $config) ? $config['bin-dir'] : 'bin';
+              if(array_key_exists("bin-dir", $config))
+                  $binDir = $config["bin-dir"];
+              else
+                  $binDir = "bin";
 
-          if (array_key_exists('bin', $config)) {
-              if (!file_exists('vendor/'.$binDir)) {
-                  mkdir('vendor/'.$binDir);
-              }
+              if(array_key_exists("bin", $config))
+              {
+                  if(!file_exists("vendor/".$binDir))
+                      mkdir("vendor/".$binDir);
 
-              foreach ($config['bin'] as $bin) {
-                  symlink('../../'.$bin, 'vendor/'.$binDir.'/'.basename($bin));
+                  foreach($config["bin"] as $bin)
+                      symlink("../../".$bin, "vendor/".$binDir."/".basename($bin));
               }
           }
+          ?>
         '';
       };
 
@@ -120,7 +167,7 @@ let
       extraArgs = removeAttrs args [ "name" "packages" "devPackages" "buildInputs" ];
     in
     stdenv.mkDerivation ({
-      pname = "composer-${pname}";
+      name = "composer-${name}";
       buildInputs = [ php composer ] ++ buildInputs;
 
       inherit unpackPhase buildPhase;
@@ -215,7 +262,8 @@ let
         runHook postInstall
     '';
   } // extraArgs);
-in {
+in
+{
   composer = stdenv.lib.makeOverridable composer;
   buildZipPackage = stdenv.lib.makeOverridable buildZipPackage;
   buildPackage = stdenv.lib.makeOverridable buildPackage;
```